### PR TITLE
Apply `black` code formatter to the python codebase

### PR DESCRIPTION
**Pull request by youtux**
### 🤔 What's changed?
* Apply `black` code formatter to the python codebase, to enforce a standard formatting
* Add a `.pre-commit-config.yaml` to the project, so that we can enforce this convention
* TODOs
  - [ ] ~~add github action to check the pre-commit hooks have been applied~~ Will add [pre-commit.ci](https://pre-commit.ci) app once this lands
  - [x] merge https://github.com/cucumber/gherkin/pull/283 first

### ⚡️ What's your motivation? 

The python codebase does not really use a standard formatting, this is arguably widely accepted as "annoying".

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://github.com/cucumber/.github/tree/main?tab=coc-ov-file)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Review by mpkorstanje — COMMENTED** [review]
I see that you are adding pre-commit hooks. Are these setup in such a way that it will not interfere with contributors who make contributions without python installed?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  - repo: https://github.com/psf/black-pre-commit-mirror
+    rev: 24.8.0
+    hooks:
+      - id: black
+        # It is recommended to specify the latest version of Python
+        # supported by your project here, or alternatively use
+        # pre-commit's default_language_version, see
+        # https://pre-commit.com/#top_level-default_language_version
+        language_version: python3.8
+        exclude: "python/gherkin/parser.py"
```
Comment (mpkorstanje) [inline .pre-commit-config.yaml:14]: Ideally this is kept in the `python` folder.

Comment (youtux) [reply .pre-commit-config.yaml]: done in f1e0e7e2

**Comment by youtux** [comment]
> I see that you are adding pre-commit hooks. Are these setup in such a way that they will not interfere with contributors who make contributions without python installed?

yes

**Review by jsa34 — APPROVED** [review]
Looks good to me

**Comment by jsa34** [comment]
I'm assuming the GitHub actions work will be in separate PR

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
```
Comment (mpkorstanje) [inline .pre-commit-config.yaml:3]: Would it be possible to keep this whole file in the `python` folder?

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
I'm not sure about the pre-commit hook yet. Looks like it will activate regardless.

**Comment by youtux** [comment]
> One thing to bare in mind (I'm doing the same in ruby), is some long methods in gherkin are written in a way that is almost language agnostic, so it reads the same in every language.

not sure how this applies in this PR @luke-hill

File: .pre-commit-config.yaml (hunk 2 above)
Comment (youtux) [reply .pre-commit-config.yaml:3]: pre-commit installs hooks in a completely isolated environment that does not affect global packages. If other contributors don't touch the `python` folder, the `black` hook won't even run at all.

Comment (youtux) [reply .pre-commit-config.yaml:3]: I tried putting `.pre-commit-config.yaml` under the `python` directory, but it won't allow any other `.pre-commit-config.yaml` to be used if anyone decides to have one under `java/` or any other folder.
Seems like a limitation of pre-commit.
So the only way is to keep it on the root dir, and explicitly whitelist the dirs that each hook should handle.

**Pull request merged**